src/app/query/bitcoin/address/utxos-by-address.hooks.ts

```diff
@@ -0,0 +1,65 @@
+import { useMemo } from 'react';
+
+import { useCurrentAccountNativeSegwitIndexZeroSigner } from '@app/store/accounts/blockchain/bitcoin/native-segwit-account.hooks';
+
+import { useInscriptionByAddressQuery } from '../ordinals/use-inscriptions.query';
+import { useBitcoinPendingTransactionsInputs } from './transactions-by-address.hooks';
+import { useGetUtxosByAddressQuery } from './utxos-by-address.query';
+
+/**
+ * Warning: ⚠️ These are **all** UTXOs, including Stamped and Inscribed UTXOs.
+ * You should probably use `useSpendableCurrentNativeSegwitAccountUtxos` instead.
+ */
+export function useCurrentNativeSegwitUtxos() {
+  const nativeSegwitSigner = useCurrentAccountNativeSegwitIndexZeroSigner();
+  return useGetUtxosByAddressQuery(nativeSegwitSigner.address);
+}
+
+export function useAllSpendableNativeSegwitUtxos(address: string) {
+  const {
+    data: inscriptions,
+    hasNextPage: hasMoreInscriptionsToLoad,
+    isLoading: isLoadingInscriptions,
+  } = useInscriptionByAddressQuery(address);
+
+  return useGetUtxosByAddressQuery(address, {
+    select(utxos) {
+      // While infinite query check has more data to load, or Stamps are loading
+      // assume nothing is spendable
+      if (hasMoreInscriptionsToLoad || isLoadingInscriptions) return [];
+
+      const inscribedUtxos = inscriptions?.pages.flatMap(page => page.results) ?? [];
+
+      return utxos.filter(
+        utxo =>
+          !inscribedUtxos.some(
+            inscription =>
+              utxo.txid === inscription.tx_id && utxo.vout === Number(inscription.offset)
+          )
+      );
+    },
+  });
+}
+
+// Safety check to make sure we don't reuse utxos in a pending tx
+function useSpendableAndNotPendingNativeSegwitUtxos(address: string) {
+  const { data: utxos = [], refetch } = useAllSpendableNativeSegwitUtxos(address);
+  const pendingInputs = useBitcoinPendingTransactionsInputs(address);
+
+  return useMemo(() => {
+    return {
+      utxos: utxos.filter(
+        utxo =>
+          !pendingInputs.find(
+            input => input.prevout.scriptpubkey_address === address && input.txid === utxo.txid
+          )
+      ),
+      refetch,
+    };
+  }, [address, pendingInputs, refetch, utxos]);
+}
```

Comment:
Filtering logic is tricky and being able to compose this easily will help us if more filtering requirements come along (and thoughts on how to not lose the React Query props).

We need to reuse the filtering logic of the `select` in `useAllSpendableNativeSegwitUtxos` in `useSpendableAndNotPendingNativeSegwitUtxos`. To do this you're using the hook and filtering with a `useMemo` (as we can't pass a second `select`). This works, but loses the query props that might be useful elsewhere.

Another approach here, might be to move the logic of `select` into it's own function, and composing the logic of `useSpendableAndNotPendingNativeSegwitUtxos` with that new filtering function, and the pending tx filter, and passing that to a new `select` function.

To clean this up further, and make things more composable, what if the filtering logic belonged in its own hook?


```ts
function useFilterAddressNativeSegwitInscriptions(address: string) {}
function useFilterAddressNativeSegwitPendingTransactions(address: string) {}

export function useAllSpendableNativeSegwitUtxos(address: string) {
  const filterOutInscriptions = useFilterAddressNativeSegwitInscriptions(address);
  return useGetUtxosByAddressQuery(address, {
    select(utxos) {
      return filterOutInscriptions(utxos);
    },
  });
}

export function useSpendableAndNotPendingNativeSegwitUtxos(address: string) {
  const filterOutInscriptions = useFilterAddressNativeSegwitInscriptions(address);
  const filterPendingTxs = useFilterAddressNativeSegwitPendingTransactions(address);
  return useGetUtxosByAddressQuery(address, {
    select(utxos) {
      return filterPendingTxs(filterOutInscriptions(utxos));
    },
  });
}
```

Reply:
Yeah, this is great. I'll refactor it today.